psycopg/psycopg/cursor.py

```diff
@@ -502,6 +514,47 @@ def _scroll(self, value: int, mode: str) -> None:
             raise IndexError("position out of bound")
         self._pos = newpos
 
+    def _fetch_pipeline_gen(self) -> PQGen[None]:
```

Comment:
In this method, if I understand correctly, when a cursor receives an execute*() call, it will go in the connection queue, look for itself in the cursors to wait, dispatching previously received results to the respective other cursors, and then proceed on.

It doesn't feel the most straightforward implementation to me. There is the initial check if the cursor is in the pipeline which is O(n) (in a likely small n, sure). But the entire idea of getting the result to be driven by the user seems backwards compared to the flow of information: who is in charge here is the connection and the results it receives.

The way I was thinking about this algorithm is more:

- the connection has a queue of cursors that have executed a statement
- the cursor has a `queued` [`threading.Event`](https://docs.python.org/3/library/threading.html#event-objects). This object has `clear()` and `set()` methods to change its state and a `wait()` method. If `wait()` is called when the event is cleared, it will block, until another thread calls `set()`. If it waits when set, it doesn't block.

Things that happen:

- on cursor.execute():
  - the connection executes the query
  - the connection puts the cursor in the queue
  - `cursor.queued.clear()` is called.

- on cursor.fetch()
  - the cursor `wait()` on the queued event, meaning that it will blocked if its result hasn't arrived yet, otherwise will waste no time
  - the cursor processes the result

- when the connection receives the result:
  - it takes the first item from the cursors queue
  - sets the result on the cursor
  - calls `set()` the queued event to release a thread eventually stuck on a fetch() or to mark it ready to fetch if the client hasn't requested a result yet.

I feel this organisation more natural, with the connection "dealing" results as they come and every cursor only in charge of its own state, not changing the state of the connection and of other cursors.

What do you think?